radix-engine-stores/src/hash_tree/mod.rs

```diff
@@ -20,55 +23,289 @@ mod test;
+/// Returns the hash of the newly-created root (i.e. representing state at version
+/// `current_version` + 1).
 ///
 /// # Panics
-/// Panics if a root node for `current_version` does not exist. The caller
-/// should use `None` to denote an empty, initial state of the tree (i.e.
-/// inserting at version 1).
+/// Panics if a root node for `current_version` does not exist. The caller should use `None` to
+/// denote an empty, initial state of the tree (i.e. inserting at version 1).
 pub fn put_at_next_version<S: TreeStore>(
     store: &mut S,
     current_version: Option<Version>,
-    changes: &[(SubstateId, Option<Hash>)],
+    changes: Vec<SubstateHashChange>,
 ) -> Hash {
-    let value_set: Vec<(Hash, Option<(Hash, SubstateId)>)> = changes
-        .iter()
-        .map(|(id, value_hash)| {
-            (
-                hash(scrypto_encode(id).unwrap()),
-                value_hash.map(|value_hash| (value_hash, id.clone())),
-            )
+    let changes_by_re_node_module = index_by_re_node_module(changes);
+    let mut nested_root_changes = Vec::new();
+    for (re_node_module, substate_changes) in changes_by_re_node_module {
+        let nested_root =
+            put_substate_changes(store, current_version, &re_node_module, substate_changes);
+        nested_root_changes.push(IdChange::new(re_node_module, nested_root));
+    }
+    put_changes(
+        store,
+        current_version,
+        current_version.unwrap_or(0) + 1,
+        nested_root_changes
+            .into_iter()
+            .map(|change| to_re_node_change(change))
+            .collect(),
+    )
+}
+
+// only internals below
+
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, Hash, ScryptoCategorize, ScryptoEncode, ScryptoDecode,
+)]
+struct ReNodeModule {
+    re_node_id: RENodeId,
+    node_module_id: NodeModuleId,
+}
+
+impl ReNodeModule {
+    fn new(re_node_id: RENodeId, node_module_id: NodeModuleId) -> Self {
+        Self {
+            re_node_id,
+            node_module_id,
+        }
+    }
+}
+
+fn index_by_re_node_module(
+    changes: Vec<SubstateHashChange>,
+) -> HashMap<ReNodeModule, Vec<IdChange<SubstateOffset, Hash>>> {
+    let mut by_re_node_module = HashMap::<ReNodeModule, Vec<IdChange<SubstateOffset, Hash>>>::new();
+    for change in changes {
+        let substate_id = change.id;
+        by_re_node_module
+            .entry(ReNodeModule::new(substate_id.0, substate_id.1))
+            .or_insert_with(|| Vec::new())
+            .push(IdChange::new(substate_id.2, change.changed));
+    }
+    by_re_node_module
+}
+
+struct TreeRoot<P> {
+    hash: Hash,
+    node: TreeNode<P>,
+}
+
+fn put_substate_changes<S: TreeStore>(
+    store: &mut S,
+    current_version: Option<Version>,
+    re_node_module: &ReNodeModule,
+    changes: Vec<IdChange<SubstateOffset, Hash>>,
+) -> Option<TreeRoot<SubstateOffset>> {
+    let substate_entry = get_substate_entry(store, current_version, re_node_module);
```

Comment:
I don't understand the name `substate_entry` isn't this more `get_leaf_in_re_node_module_tree`?

Reply:
you are right, `substate_entry` is ambiguous; let's go for `get_re_node_module_leaf_entry` (because I am getting the leaf node + its version = "an entry").